### PR TITLE
RUST-1400 Implement upload and download methods for GridFS

### PR DESCRIPTION
**Pull request by sanav33**

### CODE REVIEW

**Review by isabelatkinson — COMMENTED** [review]
looking good! these comments are just for the uploading code, still need to review the download methods

File: src/gridfs.rs [hunk 1]
```diff
@@ -131,30 +178,78 @@ impl tokio::io::AsyncRead for GridFsDownloadStream {
...
     }
 
     /// Gets the selection criteria of the [`GridFsBucket`].
     pub fn selection_criteria(&self) -> Option<&SelectionCriteria> {
-        self.options.selection_criteria.as_ref()
+        self.inner.options.selection_criteria.as_ref()
+    }
+
+    /// Gets a handle to the files collection for the [`GridFsBucket`].
+    pub fn files(&self) -> &Collection<FilesCollectionDocument> {
```
[the remark this answers is not in the document]
Comment (sanav33) [reply src/gridfs.rs:236]: Yeah this makes sense. These were included as convenience methods to access the `inner` fields for the I/O methods. Don't see any reason to expose them.

[the remark this answers is not in the document]
Comment (sanav33) [reply src/gridfs.rs:236]: Yeah, that makes sense. I don't think they need to be exposed.

File: src/gridfs.rs [hunk 2]
```diff
@@ -185,36 +292,115 @@ impl GridFsBucket {
...
+                            .build();
+                        chunks
+                            .delete_many(doc! { "files_id": &files_id }, delete_options)
+                            .await?;
+                        return Err(ErrorKind::Io(Arc::new(e)).into());
+                    }
+                };
+                curr_length += bytes_read;
+                if bytes_read == 0 {
+                    break 'outer;
```
[the remark this answers is not in the document]
Comment (sanav33) [reply src/gridfs.rs:344]: This looks like an error. Thanks for the catch! Fixed.

**Comment by isabelatkinson** [comment]
closing out for now; going to create a new PR with these commits and unified test runner support

**Pull request closed without merging**